apps/manager_app/device_authentication.c

```diff
@@ -0,0 +1,274 @@
+    // TODO: Handle proto decode error code
+    usb_clear_event();
+    return;
+  }
+
+  /* Decode recieved request_type */
+  pb_size_t request_type = get_request_type(
+      (const manager_auth_device_request_t *)&(query.auth_device));
+
+  if (SIGN_RANDOM_NUM == state) {
+    if (MANAGER_AUTH_DEVICE_REQUEST_CHALLENGE_TAG == request_type) {
+      /* Fetch and encode random challenge signature, send it to the host and
+       * wait for further USB requests. */
+      manager_result_t result =
+          get_manager_result_template(MANAGER_RESULT_AUTH_DEVICE_TAG);
+      result.auth_device =
+          sign_random_challenge(&(query.auth_device.challenge.challenge[0]));
+
+      uint8_t response[DEVICE_AUTH_RESPONSE_SIZE] = {0};
+      ASSERT(encode_and_send_manager_result(
+          &result, &response[0], sizeof(response)));
+      *state_ptr = RESULT;
+    } else if (MANAGER_AUTH_DEVICE_REQUEST_RESULT_TAG == request_type) {
+      /* If SIGN_RANDOM_NUM == state, but the request_type received is
+       * MANAGER_AUTH_DEVICE_REQUEST_RESULT_TAG, it's an unexpected step in
+       * the flow. The device will treat it as an attempt to force device
+       * authentication status */
+      device_auth_handle_response(false);
+      send_flow_complete();
+      *state_ptr = FLOW_COMPLETE;
+    } else {
+      // TODO: Handle error scenario in which an unexpected request_type was
+      // detected
+      usb_clear_event();
+    }
+  } else if (RESULT == state) {
+    bool verified = query.auth_device.result.verified;
+    device_auth_handle_response(verified);
+    send_flow_complete();
+    *state_ptr = FLOW_COMPLETE;
+  }
+
+  return;
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+void device_authentication_flow(const manager_query_t *query) {
+  if (false == check_manager_request(query, MANAGER_QUERY_AUTH_DEVICE_TAG)) {
+    // TODO: Handle bad data error
+    // TODO: Early flow exit
+  }
+
+  pb_size_t request_type = get_request_type(&(query->auth_device));
+  if (MANAGER_AUTH_DEVICE_REQUEST_INITIATE_TAG != request_type) {
+    // TODO: Handle error scenario in which an unexpected request_type was
+    // detected
+    // TODO: Early flow exit
+  }
+
+  // TODO: Check if it's a forced device authentication, in which case we
+  // will take users permission to perform authentication again
+
+  /* Set flow status: Dummy for now! */
+  core_status_set_flow_status(MANAGER_AUTH_DEVICE_STATUS_USER_CONFIRMED);
+
+  delay_scr_init(ui_text_message_device_authenticating, 100);
+
+  /* First state of the device authentication would be SIGN_SERIAL_NUMBER */
+  device_auth_flow_states_e state = SIGN_SERIAL_NUM;
+  send_serial_signature(&state);
+
+  while (FLOW_COMPLETE != state) {
+    evt_status_t event = get_events(EVENT_CONFIG_USB, MAX_INACTIVITY_TIMEOUT);
+    if (true == event.p0_event.flag) {
+      // TODO
```

Comment:
if it is straight forward, let's add the handling in this PR itself.

Reply:
added for now, but we need to make common p0 handlers soon